File: Data Structures & Algorithms/subsets/submission-1.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> subset = new ArrayList<>();
        helper(nums, 0, res, subset);
        return res;
    }
    private void helper(int[] nums, int i, List<List<Integer>> res, List<Integer> subset){
        if (i == nums.length){
            res.add(new ArrayList<>(subset));
            return;
        }
        // with i
        subset.add(nums[i]);
        helper(nums, i+1, res, subset);
        // without i
        subset.remove(subset.size() - 1);
        helper(nums, i+1, res, subset);
    }
}

File: Data Structures & Algorithms/permutations/submission-1.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        res = []
        def helper(subset, mask):
            if len(subset) == n:
                res.append(subset.copy())
                return
            
            for i in range(n):
                if mask & (1 << i):
                    continue
                subset.append(nums[i])
                helper(subset, mask | (1 << i))
                subset.pop()
        
        helper([], 0)
        return res
